clash_lib/src/common/io.rs

```diff
@@ -1,291 +1,383 @@
+//                         }
+//                         Poll::Ready(Err(err)) => return Poll::Ready(Err(err)),
+//                         Poll::Pending => (),
+//                     }
+//                 }
+//                 TransferState::Done => (),
+//             }
 
-    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        // Unpack self into mut refs to each field to avoid borrow check issues.
-        let CopyBidirectional {
-            a,
-            b,
-            a_to_b,
-            b_to_a,
-            a_to_b_count,
-            b_to_a_count,
-            a_to_b_delay,
-            b_to_a_delay,
-            a_to_b_timeout_duration,
-            b_to_a_timeout_duration,
-        } = &mut *self;
+//             match (&a_to_b, &b_to_a) {
+//                 (TransferState::Done, TransferState::Done) => break,
+//                 _ => return Poll::Pending,
+//             }
+//         }
 
-        let mut a = Pin::new(a);
-        let mut b = Pin::new(b);
+//         Poll::Ready(Ok((*a_to_b_count, *b_to_a_count)))
+//     }
+// }
 
+// pub async fn copy_buf_bidirectional_with_timeout<A, B>(
+//     a: &mut A,
+//     b: &mut B,
+//     size: usize,
+//     a_to_b_timeout_duration: Duration,
+//     b_to_a_timeout_duration: Duration,
+// ) -> Result<(u64, u64), std::io::Error>
+// where
+//     A: AsyncRead + AsyncWrite + Unpin + ?Sized,
+//     B: AsyncRead + AsyncWrite + Unpin + ?Sized,
+// {
+//     CopyBidirectional {
+//         a,
+//         b,
+//         a_to_b: TransferState::Running(CopyBuffer::new_with_capacity(size)?),
+//         b_to_a: TransferState::Running(CopyBuffer::new_with_capacity(size)?),
+//         a_to_b_count: 0,
+//         b_to_a_count: 0,
+//         a_to_b_delay: None,
+//         b_to_a_delay: None,
+//         a_to_b_timeout_duration,
+//         b_to_a_timeout_duration,
+//     }
+//     .await
+// }
+
+pub async fn copy_buf_bidirectional_with_timeout<A, B>(
+    a: A,
+    b: B,
+    size: usize,
+    a_to_b_timeout_duration: Duration,
+    b_to_a_timeout_duration: Duration,
+) -> Result<(u64, u64), std::io::Error>
+where
+    A: AsyncRead + AsyncWrite + Unpin + Sized + Send + 'static,
+    B: AsyncRead + AsyncWrite + Unpin + Sized + Send + 'static,
+{
+    let (mut a_reader, mut a_writer) = tokio::io::split(a);
+    let (mut b_reader, mut b_writer) = tokio::io::split(b);
+    let a_to_b = tokio::spawn(async move {
+        let mut upload_total_size = 0;
+        let mut buf = Vec::new();
+        buf.resize(size, 0);
         loop {
-            match a_to_b {
-                TransferState::Running(buf) => {
-                    let res = buf.poll_copy(cx, a.as_mut(), b.as_mut());
-                    match res {
-                        Poll::Ready(Ok(count)) => {
-                            *a_to_b = TransferState::ShuttingDown(count);
+            match tokio::time::timeout(a_to_b_timeout_duration, a_reader.read(&mut buf[..])).await {
```

Comment:
have you tested this?

does this not break the long running connections ?

Reply:
As exposed by the signature `copy_buf_bidirectional_with_timeout`, which is specified by `timeout`, and this function is only called in `dispatch_stream` in https://github.com/Watfaq/clash-rs/blob/master/clash_lib/src/app/dispatcher/dispatcher.rs#L138, which specifies the `Duration::from_secs(10)`, so this is not for long-running connection. The current implementation does work for me.

Reply:
i think you are confused with the timeout. it's mean to be the connection idle timeout.

Reply:
can you test if your change can successfully request `https://httpbin.yba.dev/drip\?duration\=11` ?

Reply:
I am not confused by timeout`, [tokio::time::timeout](https://teaclave.apache.org/api-docs/crates-app/tokio/time/fn.timeout.html) does mean the read should read data within the specified time duration.

Reply:
here we need the idle timeout, not to timeout for a long entire io duration.

Reply:
> here we need the idle timeout, not to timeout for a long entire io duration.

[tokio::time::timeout](https://teaclave.apache.org/api-docs/crates-app/tokio/time/fn.timeout.html) exactly is the idle timeout.

Reply:
can you test if your change can successfully request https://httpbin.yba.dev/drip\?duration\=11 ?

Reply:
> can you test if your change can successfully request [https://httpbin.yba.dev/drip\?duration\=11](https://httpbin.yba.dev/drip%5C?duration%5C=11) ?

I tried to use the real `config.yaml` to start the program, which gives the error 
````
ERROR clash_lib: panic hook: PanicInfo { payload: Any { .. }, message: Some(called `Result::unwrap()` on an `Err` value: InvalidConfig("unsupported network: grpc")), location: Location { file: "clash/src/main.rs", line: 68, col: 6 }, can_unwind: true, force_no_backtrace: false }
````

But, this ` InvalidConfig("unsupported network: grpc"))` is not caused by this change.

Reply:
try start with https://github.com/Watfaq/clash-rs/blob/master/clash/tests/data/config/simple.yaml

Reply:
> try start with https://github.com/Watfaq/clash-rs/blob/master/clash/tests/data/config/simple.yaml

I have successfully started the program with this `simple.yaml`, then what do I need to do for the next step?

Reply:
@ibigbug As shown in the snapshot, this is ok.
![855117179](https://github.com/Watfaq/clash-rs/assets/17973891/2d14ccf4-4767-4aff-915d-0bb797052c7f)

Reply:
i'm talking about this url `https://httpbin.yba.dev/drip\?duration\=11`

Reply:
@ibigbug Response `curl -x http://127.0.0.1:8080 https://httpbin.yba.dev/drip\?duration\=11
**********%  `

Reply:
![1812087872](https://github.com/Watfaq/clash-rs/assets/17973891/6af3140f-5924-4d8d-b1c1-571b64cdabc9)

It can also response curl: (56) Failure when receiving data from the peer, I suppose this is idle timeout? We may set the Duration longer than `10`.

Reply:
right, 11 might be too closed to the timeout.

if you can try with duration = 15 or some longer value you should see some error like this

```
curl: (56) Failure when receiving data from the peer
```

as it closes the conn premaruely

```
-> % curl -v -k -x socks5h://127.0.0.1:8889 https://httpbin.yba.dev/drip\?duration\=13
*   Trying 127.0.0.1:8889...
* Connected to 127.0.0.1 (127.0.0.1) port 8889 (#0)
* SOCKS5 connect to httpbin.yba.dev:443 (remotely resolved)
* SOCKS5 request granted.
* Connected to 127.0.0.1 (127.0.0.1) port 8889 (#0)
* ALPN: offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-CHACHA20-POLY1305-SHA256
* ALPN: server accepted h2
* Server certificate:
*  subject: C=US; ST=California; L=San Francisco; O=Cloudflare, Inc.; CN=sni.cloudflaressl.com
*  start date: Feb 10 00:00:00 2023 GMT
*  expire date: Feb  9 23:59:59 2024 GMT
*  issuer: C=US; O=Cloudflare, Inc.; CN=Cloudflare Inc ECC CA-3
*  SSL certificate verify ok.
* using HTTP/2
* h2 [:method: GET]
* h2 [:scheme: https]
* h2 [:authority: httpbin.yba.dev]
* h2 [:path: /drip?duration=13]
* h2 [user-agent: curl/8.1.2]
* h2 [accept: */*]
* Using Stream ID: 1 (easy handle 0x14480bc00)
> GET /drip?duration=13 HTTP/2
> Host: httpbin.yba.dev
> User-Agent: curl/8.1.2
> Accept: */*
>
* Closing connection 0
curl: (56) Failure when receiving data from the peer

```

Reply:
it's not about the Duration value, you are changing the timeout behaviour.

I think this PR causes error. I'll not further proceed this.

Reply:
that's exactly the point.

sorry I got something else to fix, I'll not reply to this thread.

Reply:
What the "connection idle-timeout" do you mean here? My understanding is if the connection cannot read data within `X` time, the connection is considered to be idle, the `timeout` just specifies what maximum time the connection can be in this idle state before reading.

Incidentally, in your original implementation, whatever the number is given in `https://httpbin.yba.dev/drip\?duration\=number`, it always responds `********` even though the Duration in `copy_buf_bidirectional_with_timeout` is set to be `10`.

Reply:
ok i finished the other stuff. now i'm getting back to this.

as you mentioned in the link, this timeout needs to be the duration during which *no traffic is sent*, not the entire time it takes to finish the connection.

think about you need to download a large file, it take 2hrs, but in the 2hrs, there's always data being transfered, you can't just give it a 10s timeout or whatever.

Reply:
After reviewing the original implementation carefully, the timeout logic is, that only if one direction is complete, it will set the timeout for the other direction. In other words, if two directions are both pending, there is no timeout imposed on them.